### DISCUSSION TITLE
Publicodes-loader : Intégration facilitée, découpage de code pour les règles, génération du typage des règles

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#183] [Ideas]
On laisse actuellement à chaque application le travail d'importer les règles publicodes et d'instancier le moteur avec un code du type :

```js
// webpack.config.js
loaders: {
  'règles/**/*.yaml': YamlLoader
}

// engine.js
import fichier1 from "règles/fichier1.yaml"
import fichier2 from "règles/fichier2.yaml"
import Publicodes from "publicodes"

export default new Publicodes({
  ...fichier1,
  ...fichier2,
})
```

Je vois au moins 2 problèmes avec cette approche :
1. On importe systématiquement l'ensemble des règles — même si on n'a besoin que d'un sous-ensemble
2. Chaque application doit redévelopper cette logique :
    - [mon-entreprise](https://github.com/betagouv/mon-entreprise/blob/9aa9c527b1604cbc36c52b65feacd7203a54d059/scripts/build-rules.js)
    - [nosgestesclimat](https://github.com/datagir/nosgestesclimat/blob/bfa58dd9ae6954a5758970a37664d58c7968a9dc/scripts/rulesToJSON.js)
    - [futureco](https://github.com/laem/futureco-data/blob/4bd0d45bcca5d243f46500afb82ddfa2bf0bfda7/rulesToJSON.js)
    - [code-du-travail-numérique](https://github.com/SocialGouv/code-du-travail-numerique/blob/e8edcaa32fda239890ba5f638af0c11ade22d028/packages/code-du-travail-modeles/src/build.ts)

Je propose de créer un nouveau paquet `publicodes-loader` facultatif qui permette d'importer facilement un fichier publicodes dans une application Webpack / Vite / Rollup. Je pense que cela peut être fait de manière universelle avec https://github.com/unjs/unplugin.

Au delà de faciliter l'intégration de règles publicodes dans une appli web complète, l'intérêt serait aussi de supporter du découpage de code au niveau des fichiers. L'idée serait de compiler chaque fichier `*.publicodes.yaml` en fichier ES module avec les bons imports vers les dépendances : 

```js
// salarié.publicodes.yaml
import ImpotsRules from "./impots.publicodes.yaml"
import BaseRules from "./base.publicodes.yaml"

export default {
  ...ImpotsRules,
  ...BaseRules,
  "contrat salarié" : {...},
  "contrat salarié . salaire net" : {...},
  ...
}
```

Puis pour l'utiliser 

```js
import SalariéRules from "règles/salarié.publicodes.yaml"
import Publicodes from "publicodes"

export default new Publicodes(SalariéRules)
```

Il serait peut-être même souhaitable d'aller un cran plus loin et que `publicodes-loader` expose directement une instance du moteur sans que l'utilisateur ait a importer les règles d'une part, et le moteur d'autre part, avant de ré-exporter une instance du moteur.

Ce paquet `publicodes-loader` pourra aussi générer les fichiers d'interface Typescript avec notamment la liste des noms des règles ainsi que le type de chaque règle (numérique, booléen, texte, nullable), cf. https://github.com/betagouv/mon-entreprise/issues/1995

**Comment by johangirod** [comment]
Reliés : 
- #86 
- #89

On pourrait aussi imaginer pouvoir importer des fichiers `.csv` avec `publicodes-loader` : 
- #88
